Common++/header/PointerVector.h

```diff
@@ -55,22 +66,23 @@ namespace pcpp
 		 */
 		PointerVector(const PointerVector& other)
 		{
-			try
+			try 
 			{
+				std::lock_guard<Mutex> lk(other.m_Mutex);
 				for (const auto iter : other)
-				{
-					T* objCopy = new T(*iter);
-					try
 					{
-						m_Vector.push_back(objCopy);
+						T* objCopy = new T(*iter);
+						try
+						{
+							m_Vector.push_back(objCopy);
+						}
+						catch (const std::exception&)
+						{
+							delete objCopy;
+							throw;
+						}
 					}
-					catch (const std::exception&)
-					{
-						delete objCopy;
-						throw;
-					}
-				}
-			}
+			} 
 			catch (const std::exception&)
 			{
```

Comment:
we may need a lock here?

Reply:
@tigercosmos in the catch block?

Reply:
That should not be an issue as this is a copy constructor. So nothing should have a reference to the object under construction yet. The main concern is locking the object that is being copied from as that could have other references.

Reply:
I see, make sense. About the lock in the try block, do you think it's more efficient to put at the outer one or the inner one; in other orders, locking the whole list or one by one is better?

Reply:
Locking and unlocking once is generally better, as it would introduce less performance overhead. Generally we are going though the entire list anyway, IMO, locking the whole list once is the better option.

Reply:
I also have a concern locking one by one will increase the overhead instead. Let's keep the current way.

Reply:
Another thing is that locking/unlocking for each element induces the potential of the list being modified in the middle of the copy operation. Thus making both approaches inequivalent from data consistency perspective.